beacon/sync/src/main/java/tech/pegasys/teku/beacon/sync/historical/ReconstructHistoricalStatesService.java

```diff
@@ -13,18 +13,120 @@
 
+import java.io.IOException;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.exceptions.InvalidConfigurationException;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.service.serviceutils.Service;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.state.Checkpoint;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.datastructures.util.ChainDataLoader;
+import tech.pegasys.teku.storage.api.StorageUpdateChannel;
+import tech.pegasys.teku.storage.client.CombinedChainDataClient;
 
 public class ReconstructHistoricalStatesService extends Service {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final CombinedChainDataClient chainDataClient;
+  private final Spec spec;
+  private final Optional<String> genesisStateResource;
+  private final StorageUpdateChannel storageUpdateChannel;
+
+  public ReconstructHistoricalStatesService(
+      final StorageUpdateChannel storageUpdateChannel,
+      final CombinedChainDataClient chainDataClient,
+      final Spec spec,
+      final Optional<String> genesisStateResource) {
+    this.storageUpdateChannel = storageUpdateChannel;
+    this.chainDataClient = chainDataClient;
+    this.spec = spec;
+    this.genesisStateResource = genesisStateResource;
+  }
 
   @Override
   protected SafeFuture<?> doStart() {
-    return SafeFuture.COMPLETE;
+    if (genesisStateResource.isEmpty()) {
+      return SafeFuture.failedFuture(
+          new IllegalStateException("Genesis state resource not provided"));
+    }
+
+    final String resource = genesisStateResource.get();
+    final BeaconState genesisState;
+    try {
+      genesisState = ChainDataLoader.loadState(spec, resource);
+    } catch (IOException e) {
+      LOG.error("Failed to load initial state", e);
+      return SafeFuture.failedFuture(
+          new InvalidConfigurationException(
+              "Failed to load initial state from " + resource + ": " + e.getMessage()));
+    }
+
+    final SafeFuture<Optional<Checkpoint>> future = chainDataClient.getInitialAnchor();
+    future
+        .thenAccept(
+            checkpoint -> {
+              if (checkpoint.isEmpty()) {
+                return; // todo confirm skip, or should I throw exception
+              }
+
+              applyBlocks(genesisState, checkpoint.get().getEpochStartSlot(spec));
+            })
+        .finish(LOG::error); // todo ignores return value
+
+    return SafeFuture.COMPLETE; // todo check
+  }
+
+  public void applyBlocks(final BeaconState genesisState, final UInt64 anchorSlot) {
+    Context context = new Context(genesisState, UInt64.ONE, anchorSlot); // todo correct slot 1?
+    applyNextBlock(context).finish(LOG::error); // todo ignores return value
```

Comment:
This probably needs to use `.finish` but would be better to have the error go to the console - we've been unable to reconstruct historic states before of a failure.  We probably should get it to retry after a period - it may have failed because the disk was full so if it tries again later then it can resume making progress if disk space has become available. Probably best to do as a follow up change though.